storage/internal/benchmarks/client_pool.go

```diff
@@ -31,130 +30,125 @@ import (
-// and is the intended way to use Storage clients.
-//
-// There is no limit to how many clients will be created, but it should be around
-// the order of 5 * min(workers, max_samples).
+// clientPool pools a number of Storage clients for use without blocking, ie.
+// a client that is received through Get may still be in use by one or more other
+// calls.
 type clientPool struct {
-	New     func() *storage.Client
-	clients []*storage.Client
+	clients     []*storage.Client
+	clientQueue chan *storage.Client
 }
 
-func (p *clientPool) Get() *storage.Client {
-	// Create the slice if not already created
-	if p.clients == nil {
-		p.clients = make([]*storage.Client, 0)
+// newClientPool initializes the pool with numClients clients initialized using
+// the initializeClient func.
+// Returns the client pool and a cleanup func to be called to close the pool.
+func newClientPool(initializeClient func() (*storage.Client, error), numClients int) (*clientPool, func()) {
+	p := &clientPool{
+		clients:     make([]*storage.Client, numClients),
+		clientQueue: make(chan *storage.Client, numClients),
 	}
 
-	// If there is an unused client, return it
-	if len(p.clients) > 0 {
-		c := p.clients[0]
-		p.clients = p.clients[1:]
-		return c
-	}
+	for i := 0; i < numClients; i++ {
+		var err error
+		p.clients[i], err = initializeClient()
+		if err != nil {
+			log.Fatalf("initializeClient: %v", err)
+		}
 
-	// Otherwise, create a new client and return it
-	return p.New()
-}
+		// Fill the queue with clients as they are created
+		p.clientQueue <- p.clients[i]
+	}
 
-func (p *clientPool) Put(c *storage.Client) {
-	p.clients = append(p.clients, c)
+	return p, func() {
+		for _, c := range p.clients {
+			c.Close()
+		}
+	}
 }
 
-// we can share clients as long as the app buffer sizes are constant
-var httpClients, gRPCClients *clientPool
+// Get rotates through clients. This means the work may not be evenly distributed,
+// particularly if using varying object sizes.
+func (p *clientPool) Get() *storage.Client {
+	client := <-p.clientQueue
 
-var nonBenchmarkingClients = clientPool{
-	New: func() *storage.Client {
-		// For debuggability's sake, these are HTTP
-		clientMu.Lock()
-		client, err := storage.NewClient(context.Background())
-		clientMu.Unlock()
-		if err != nil {
-			log.Fatalf("storage.NewClient: %v", err)
-		}
+	// return client to queue so that it will be used again without blocking
+	p.clientQueue <- client
 
-		return client
-	},
+	return client
 }
 
-func initializeClientPools(opts *benchmarkOptions) func() {
-	httpClients = &clientPool{
-		New: func() *storage.Client {
-			client, err := initializeHTTPClient(context.Background(), opts.writeBufferSize, opts.readBufferSize, true)
-			if err != nil {
-				log.Fatalf("initializeHTTPClient: %v", err)
-			}
+var httpClients, gRPCClients, nonBenchmarkingClients *clientPool
 
-			return client
-		},
-	}
+func initializeClientPools(ctx context.Context, opts *benchmarkOptions) func() {
```

Comment:
Maybe add a comment to explicitly note that this creates separate pools for HTTP and gRPC if need be